File: apps/clojure/src/algorithm/strings.clj
(ns algorithm.strings
  "第7章 文字列処理")

(defn bf-match
  "ブルートフォース文字列探索"
  [txt pat]
  (let [n (count txt)
        m (count pat)]
    (if (zero? m)
      0
      (loop [i 0]
        (if (> i (- n m))
          -1
          (let [matched (loop [j 0]
                          (cond
                            (>= j m) true
                            (not= (nth txt (+ i j)) (nth pat j)) false
                            :else (recur (inc j))))]
            (if matched i (recur (inc i)))))))))

(defn- build-kmp-table
  "KMP 法の失敗関数テーブルを構築する"
  [pat]
  (let [m (count pat)
        table (int-array m 0)]
    (loop [i 1 k 0]
      (when (< i m)
        (let [k (loop [k k]
                  (if (and (> k 0) (not= (nth pat k) (nth pat i)))
                    (recur (aget table (dec k)))
                    k))
              k (if (= (nth pat k) (nth pat i)) (inc k) k)]
          (aset table i k)
          (recur (inc i) k))))
    (vec table)))

(defn kmp-match
  "KMP 文字列探索"
  [txt pat]
  (let [n (count txt)
        m (count pat)]
    (if (zero? m)
      0
      (let [table (build-kmp-table pat)]
        (loop [i 0 j 0]
          (if (>= i n)
            -1
            (let [j (loop [j j]
                      (if (and (> j 0) (not= (nth txt i) (nth pat j)))
                        (recur (nth table (dec j)))
                        j))
                  j (if (= (nth txt i) (nth pat j)) (inc j) j)]
              (if (= j m)
                (- i m -1)
                (recur (inc i) j)))))))))

(defn bm-match
  "Boyer-Moore 文字列探索（Bad Character ルールのみ）"
  [txt pat]
  (let [n (count txt)
        m (count pat)]
    (if (zero? m)
      0
      (let [bad-char (into {} (map-indexed (fn [i c] [c i]) pat))]
        (loop [s 0]
          (if (> s (- n m))
            -1
            (let [j (loop [j (dec m)]
                      (cond
                        (< j 0) j
                        (= (nth pat j) (nth txt (+ s j))) (recur (dec j))
                        :else j))]
              (if (< j 0)
                s
                (let [skip (- j (get bad-char (nth txt (+ s j)) -1))]
                  (recur (+ s (max 1 skip))))))))))))

(defn count-chars
  "文字列中の各文字の出現回数を map で返す"
  [s]
  (if (empty? s)
    {}
    (frequencies s)))

(defn str-reverse
  "文字列を逆順にして返す"
  [s]
  (apply str (reverse s)))

(defn is-palindrome?
  "文字列が回文かどうかを判定する"
  [s]
  (= s (str-reverse s)))

File: apps/fsharp/Algorithm/Strings.fs
module Strings

open System
open System.Collections.Generic

/// ブルートフォース文字列探索
let bfMatch (txt: string) (pat: string) =
    let n = txt.Length
    let m = pat.Length
    if m = 0 then 0
    else
        let mutable result = -1
        let mutable i = 0
        while i <= n - m && result = -1 do
            let mutable j = 0
            while j < m && txt.[i + j] = pat.[j] do
                j <- j + 1
            if j = m then result <- i
            i <- i + 1
        result

/// KMP 文字列探索
let kmpMatch (txt: string) (pat: string) =
    let n = txt.Length
    let m = pat.Length
    if m = 0 then 0
    else
        // KMP テーブル構築
        let table = Array.zeroCreate m
        let mutable k = 0
        for i in 1..m-1 do
            while k > 0 && pat.[k] <> pat.[i] do
                k <- table.[k - 1]
            if pat.[k] = pat.[i] then k <- k + 1
            table.[i] <- k
        // 探索
        let mutable j = 0
        let mutable result = -1
        let mutable i = 0
        while i < n && result = -1 do
            while j > 0 && txt.[i] <> pat.[j] do
                j <- table.[j - 1]
            if txt.[i] = pat.[j] then j <- j + 1
            if j = m then result <- i - m + 1
            i <- i + 1
        result

/// BM 文字列探索（簡易版）
let bmMatch (txt: string) (pat: string) =
    let n = txt.Length
    let m = pat.Length
    if m = 0 then 0
    else
        let badChar = Dictionary<char, int>()
        for i in 0..m-1 do badChar.[pat.[i]] <- i
        let mutable s = 0
        let mutable result = -1
        while s <= n - m && result = -1 do
            let mutable j = m - 1
            while j >= 0 && pat.[j] = txt.[s + j] do
                j <- j - 1
            if j < 0 then result <- s
            else
                let bc = if badChar.ContainsKey(txt.[s + j]) then badChar.[txt.[s + j]] else -1
                let skip = max 1 (j - bc)
                s <- s + skip
        result

/// 文字の出現回数をカウントする
let countChars (s: string) =
    let result = Dictionary<char, int>()
    for c in s do
        let mutable count = 0
        result.TryGetValue(c, &count) |> ignore
        result.[c] <- count + 1
    result

/// 文字列を逆順にする
let reverseString (s: string) =
    let arr = s.ToCharArray()
    Array.Reverse(arr)
    new string(arr)

/// 回文判定
let isPalindrome (s: string) = s = reverseString s
